Modernize WatchServiceFileSystemWatcher and watch .env for dev mode

gsmet opened the pull request:
Fixes #41282

I can't say I'm a big fan of the fix but we will have to live with this for now.

@stuartwdouglas while working on this, I stumbled upon a small issue: if we are filtering tests, the counter of run tests (`All 5 tests are passing (0 skipped)`) is not updated when we add filters (even when we force a re-run).
My understanding is that we actually don't clear the test results but only push diffs to the test results (in `TestState`). I wonder if when we fully rerun the tests, we should actually clear the status but I'm not entirely sure, that's what you wanted?
The reproducer here is quite useful: https://github.com/flyinfish/quarkus--continuous-testing-dotenv . Run dev mode, then uncomment the filters in `src/main/resources/application.properties` and you should see the problem.

@radcortez at some point, we should probably think of a way to achieve this in a fully supported manner as this thing is quite brittle :). Not sure it's high priority though.

gsmet committed:
Include .env in the config hack of TestSupport

Fixes #41282

I can't say I'm a big fan of the fix but we will have to live with this
for now.

radcortez commented:
> @radcortez at some point, we should probably think of a way to achieve this in a fully supported manner as this thing is quite brittle :). Not sure it's high priority though.

We could probably use https://github.com/quarkusio/quarkus/blob/1d6599995fe314ceedfab52f69f90598caa416c4/core/runtime/src/main/java/io/quarkus/runtime/configuration/ConfigUtils.java#L75-L87 and retrieve any required configuration from that instance.

gsmet committed:
Modernize WatchServiceFileSystemWatcher

- Switch to java.nio consistently
- Allow to monitor a directory for specific files
- Monitor .env file

gsmet commented:
@radcortez so I went with your idea but went with an even more stripped version of SmallRyeConfig to avoid discovering the converters... could you have a look at the first commit.

@geoand I would appreciate a look at the second commit. I modernized `WatchServiceFileSystemWatcher` to use `java.nio` and more modern constructs and also added the ability to monitor a specific file (in this case, the `.env`).

geoand commented:
Seems reasonable, but CI seems to be complaining

radcortez commented:
> @radcortez so I went with your idea but went with an even more stripped version of SmallRyeConfig to avoid discovering the converters... could you have a look at the first commit?

Perfect!

gsmet commented:
Mkay. Doesn't look good. I will have a look at these particular tests...
